File: src/main/java/ar/com/flow/bankaccount/transaction/State.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.money.Money;
import lombok.RequiredArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@RequiredArgsConstructor
class State {
    private List<UndoStep> undoSteps = new ArrayList<>();

    void completed(Step completedStep, Money amount) {
        undoSteps.add(new UndoStep(completedStep, amount));
    }

    void rollback() {
        undoSteps.forEach(UndoStep::execute);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/transaction/State.kt
package ar.com.flow.bankaccount.transaction

import ar.com.flow.money.Money
import java.util.function.Consumer

internal class State {
    private val undoSteps: MutableList<UndoStep> = ArrayList()

    fun completed(completedStep: Step, amount: Money) {
        undoSteps.add(UndoStep(completedStep, amount))
    }

    fun rollback() {
        undoSteps.forEach(Consumer { obj: UndoStep -> obj.execute() })
    }
}